### ISSUE TITLE
only expose an opaque pointer for the different plans in the public API

### ISSUE THREAD

**Issue by ghisvail** [NFFT/nfft#5]
Right now, the public API exposes the whole plan structure, which increases the potential for API breakage in future updates. This is what happened when the <plan>_flags attribute has been renamed to just flags in a previous commit for instance.

In order to avoid that later, but also simplify the public API a bit, it would be better to only expose all plans as opaque pointers and hide the internal structure of each plan to the user, as it is done in FFTW actually.

The flipside of this change is that the x, f and f_hat can no longer be accessed directly and should be updated through the public APIs functions and / or assigned during initialization. I feel confident I can come up with a design for that following what is done in FFTW, but I wanted to discuss that here first.

**Comment by jenskeiner** [comment]
I had advocated for a change along these lines some time ago. I'm still supporting it.

Accessing f, x, f_hat should not be a problem if we nest two plan definitions, a public and a private one FFTW-style. So nothing would change for the end user.

**Comment by ghisvail** [comment]
Ok, I'll try something out in a different feature branch.

**Comment by ghisvail** [comment]
@jenskeiner 
I am confused, are you taking care of this, shall I give it a go or shall we work both on our sides and discuss the final API later ?

**Comment by jenskeiner** [comment]
Let me discuss this with the rest of the pack first.

**Comment by ghisvail** [comment]
Ok, I'll stand by.

**Comment by jenskeiner** [comment]
btw: The plan definitions have to change one way or the other as they rely on assertions on padding in structs. This is simply wrong.

**Comment by ghisvail** [comment]
What do you mean ?

**Comment by jenskeiner** [comment]
The shared plan members are defined via a macro,. That exposes the code to compiler optimizations in terms of padding between them (in each plan struct we might see different padding). We should do it via a base struct (which enforces a single padding scheme therein) and include that as a single member in each plan struct.

**Comment by ghisvail** [comment]
I guess you are referring to the mv_plan macro, which defines the common arrays and function ptrs to the forward and adjoint transforms

Technically, I don't see why this is an issue. The end-user should only see an opaque {nfft_, nfct_, ...}_plan pointer with which to interact via the relevant trafo, adjoint, precompute...routines. Then, the way you organize the plan internal structures should not matter anymore, does it ? I can't comment on the benefit of having a base struct vs macro internally, you probably know more than I do. My feeling is that it might make things easier for the solver struct, since you'd only need to write it once for the base struct and target all plans via a ptr cast ?

API-wise, if we follow what FFTW does, then all the ```init``` functions should return an allocated nfft_plan pointer instead of being routines,

**Comment by ghisvail** [comment]
Whilst working on my updated Python wrappers for NFFT, I gave this issue a further thinking.

Ideally, we'd want to only expose an opaque pointer, say ```nfft_plan```, to an internal structure, say ```nfft_plan_s```, whose implementation details are hidden. The user should only interact with the plan through computation routines, which is what FFTW is currently doing. An additional benefit is that the internals of the hidden plan structure can change without resulting in API breakage, which makes for more reliable development based off NFFT.

Only exposing an opaque pointer has the immediate disadvantage that none of the internal arrays, like x, f, and f_hat would be directly accessible anymore. This can be addressed by providing the arrays either as part of the ```nfft_init``` methods and / or arguments to the ```nfft_precompute```, ```nfft_trafo``` and ```nfft_adjoint```. Besides the ```nfft_init``` arrays should probably return the malloc'd plans since the definition of the plan structure would be hidden.

Another hurdle is the fact that the plan can also manage its own internal arrays via the ```MALLOC_``` flags. So we would need to abstract data assignement to x, f and f_hat, by providing appropriate getters and setters-like routine. Unless, an update to x, f, f_hat is only allowed when calling the ```nfft_precompute```, ```nfft_trafo``` or ```nfft_adjoint``` routines.

Based on these observations, here is a candidate API 

```
#include <fftw3.h>  /* provides fftw*_complex dtypes */

typedef struct nfft_plan_s *nfft_plan;  /* opaque ptr */
nfft_plan nfft_init(...);  /* one of many potential ctors */
void nfft_finalize(nfft_plan);  /* dtor */
void nfft_precompute(nfft_plan, double*);  /* precompute with update of x*/
void nfft_execute_precomputation(nfft_plan); /* precompute without update */
void nfft_trafo(nfft_plan, fftw_complex*, fftw_complex*); /* forward with update of f, f_hat */
void nfft_execute_trafo(nfft_plan);  /*forward without update*/
void nfft_adjoint(nfft_plan, fftw_complex*, fftw_complex*); /* forward with update of f, f_hat */
void nfft_execute_adjoint(nfft_plan);  /*adjoint without update*/
nfft_check(nfft_plan);  /* sanity checks */
/*optional getters and setters, may be left internal */
void nfft_set_x(nfft_plan, double*);
double *nfft_get_x(nfft_plan);
void nfft_set_f(nfft_plan, fftw_complex*);
fftw_complex *nfft_get_f(nfft_plan);
void nfft_set_f_hat(nfft_plan, fftw_complex*);
fftw_complex *nfft_get_f_hat(nfft_plan);
```

Just food for thoughts. Nothing may come out of this, just felt like sharing.

**Comment by ghisvail** [comment]
Just wanted to know whether there were plans to move this issue further?

**Comment by jenskeiner** [comment]
We're going to make the change. It's just that not much work has been done recently.

We also want to get rid of the SVN backend. Since I did not set up `authors.txt` when I pulled from SVN, author names have not been mapped correctly to the users. We should fix that as well.

I could make the required changes soon, but I might need to re-setup the entire repo. Any concerns on your side?

We should make these changes now before we get more traffic and start working on other changes.

**Comment by ghisvail** [comment]
> Any concerns on your side?

Nope. I am tracking the releases via the releases, so as long as you keep the tag format consistent, it will be fine on my end.
